# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/render_backend.rs
# Commit: 5506d1803e15
# Full Hash: 5506d1803e158df5504bc74a7818cb78dca6420c
# Author: Kris Taeleman <ktaeleman@mozilla.com>
# Date: 2019-10-16 05:20:00
# Description:
#   Bug 1585801 - Fixing crash when folder is not writable and refactored logic. r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D48999
# ==============================================================================

diff -r f24ddb55b48b -r 5506d1803e15 gfx/wr/webrender/src/render_backend.rs
--- a/gfx/wr/webrender/src/render_backend.rs	Sat Oct 12 23:50:45 2019 +0000
+++ b/gfx/wr/webrender/src/render_backend.rs	Tue Oct 15 17:03:22 2019 +0000
@@ -701,7 +701,7 @@
 
     notifier: Box<dyn RenderNotifier>,
     recorder: Option<Box<dyn ApiRecordingReceiver>>,
-    logrecorder: Option<Box<dyn ApiRecordingReceiver>>,
+    logrecorder: Option<Box<LogRecorder>>,
     sampler: Option<Box<dyn AsyncPropertySampler + Send>>,
     size_of_ops: Option<MallocSizeOfOps>,
     debug_flags: DebugFlags,
@@ -956,18 +956,6 @@
 
             status = match self.api_rx.recv() {
                 Ok(msg) => {
-                    if self.debug_flags.contains(DebugFlags::LOG_TRANSACTIONS) {
-                        if let None = self.logrecorder {
-                            let current_time = time::now_utc().to_local();
-                            let name = format!("wr-log-{}.log",
-                                current_time.strftime("%Y%m%d_%H%M%S").unwrap()
-                            );
-                            self.logrecorder = Some(Box::new(LogRecorder::new(&PathBuf::from(name))))
-                        }
-                    } else {
-                        self.logrecorder = None;
-                    }
-
                     if let Some(ref mut r) = self.logrecorder {
                         r.write_msg(frame_counter, &msg);
                     }
@@ -1180,6 +1168,21 @@
                         // before the `PublishDocument` messages sent by `load_capture`.
                         return RenderBackendStatus::Continue;
                     }
+                    DebugCommand::SetTransactionLogging(value) => {
+                        match (value, self.logrecorder.as_ref()) {
+                            (true, None) => {
+                                    let current_time = time::now_utc().to_local();
+                                    let name = format!("wr-log-{}.log",
+                                        current_time.strftime("%Y%m%d_%H%M%S").unwrap()
+                                    );
+                                    self.logrecorder = LogRecorder::new(&PathBuf::from(name));
+                            },
+                            (false, _) => self.logrecorder = None,
+                            _ => (),
+                        };
+
+                        return RenderBackendStatus::Continue;
+                    }
                     DebugCommand::ClearCaches(mask) => {
                         self.resource_cache.clear(mask);
                         return RenderBackendStatus::Continue;